torchrl/objectives/cql.py

```diff
@@ -741,54 +805,57 @@ def _loss_qvalue_v(self, tensordict: TensorDictBase) -> Tuple[Tensor, Tensor]:
         )
 
         min_qf1_loss = (
-            torch.logsumexp(cat_q1 / self.temperature, dim=1).mean()
+            torch.logsumexp(cat_q1 / self.temperature, dim=1)
             * self.min_q_weight
             * self.temperature
         )
         min_qf2_loss = (
-            torch.logsumexp(cat_q2 / self.temperature, dim=1).mean()
+            torch.logsumexp(cat_q2 / self.temperature, dim=1)
             * self.min_q_weight
             * self.temperature
         )
 
         # Subtract the log likelihood of data
-        min_qf1_loss = min_qf1_loss - q_pred[0].mean() * self.min_q_weight
-        min_qf2_loss = min_qf2_loss - q_pred[1].mean() * self.min_q_weight
-        alpha_prime_loss = 0
-        if self.with_lagrange:
-            alpha_prime = torch.clamp(
-                self.log_alpha_prime.exp(), min=0.0, max=1000000.0
-            )
-            min_qf1_loss = alpha_prime * (min_qf1_loss - self.target_action_gap)
-            min_qf2_loss = alpha_prime * (min_qf2_loss - self.target_action_gap)
+        cql_q1_loss = min_qf1_loss - pred_q1 * self.min_q_weight
+        cql_q2_loss = min_qf2_loss - pred_q2 * self.min_q_weight
 
-            alpha_prime_loss = (-min_qf1_loss - min_qf2_loss) * 0.5
+        # write cql losses in tensordict for alpha prime loss
+        tensordict.set(self.tensor_keys.cql_q1_loss, cql_q1_loss)
+        tensordict.set(self.tensor_keys.cql_q2_loss, cql_q2_loss)
 
-        q_pred = q_pred.squeeze(-1)
-        loss_qval = distance_loss(
-            q_pred,
-            target_value.expand_as(q_pred),
-            loss_function=self.loss_function,
-        )
+        return (cql_q1_loss + cql_q2_loss).mean(), {}
 
-        qf1_loss = loss_qval[0] + min_qf1_loss
-        qf2_loss = loss_qval[1] + min_qf2_loss
+    def alpha_prime_loss(self, tensordict: TensorDictBase) -> Tensor:
+        cql_q1_loss = tensordict.get(self.tensor_keys.cql_q1_loss)
+        cql_q2_loss = tensordict.get(self.tensor_keys.cql_q2_loss)
 
-        loss_qval = qf1_loss + qf2_loss
+        if cql_q1_loss is None:
+            raise KeyError(
+                f"Couldn't find the cql_q1_loss with key {self.tensor_keys.cql_q1_loss} in the input tensordict. "
+                "This could be caused by calling alpha_prime_loss method before cql_loss method."
+            )
+        if cql_q2_loss is None:
+            raise KeyError(
+                f"Couldn't find the cql_q2_loss with key {self.tensor_keys.cql_q2_loss} in the input tensordict. "
+                "This could be caused by calling alpha_prime_loss method before cql_loss method."
+            )
 
-        td_error = abs(q_pred - target_value)
+        alpha_prime = torch.clamp(self.log_alpha_prime.exp(), min=0.0, max=1000000.0)
```

Comment:
is `1000000.0` a  magic number? What is the purpose of this clamp (to understand how we can code this in a more grounded way)?
No need to clamp an exp to 0 below, we can just do clamp_max

Reply:
Indeed the number seems quite random, I took it directly from the official implementation [1](https://github.com/aviralkumar2907/CQL/blob/d67dbe9cf5d2b96e3b462b6146f249b3d6569796/d4rl/rlkit/torch/sac/cql.py#L271).
As far as I could see it in some tests the alpha_prime value can become quite big I guess this is just to make sure its not exploding. But I'll have a look at the paper if they mention something more specific.